src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -207,51 +331,78 @@ public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document
             uint docCookie;
             if (RunningDocumentTable.TryGetCookieForInitializedDocument(document.Key.Moniker, out docCookie))
             {
-                TryProcessOpenForDocCookie(docCookie);
+                TryProcessOpenForDocCookie(docCookie, cancellationToken);
             }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            RemovePendingDocumentInitializationTask(document);
         }
 
-        private void TryProcessOpenForDocCookie(uint docCookie)
+        private void TryProcessOpenForDocCookie(uint docCookie, CancellationToken cancellationToken)
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
+
             string moniker = RunningDocumentTable.GetDocumentMoniker(docCookie);
 
             IVsHierarchy hierarchy;
             uint itemid;
             RunningDocumentTable.GetDocumentHierarchyItem(docCookie, out hierarchy, out itemid);
 
-            var shimTextBuffer = RunningDocumentTable.GetDocumentData(docCookie) as IVsTextBuffer;
+            var projectsBuilder = ArrayBuilder<IVisualStudioHostProject>.GetInstance();
+            var documentKeysBuilder = ArrayBuilder<DocumentKey>.GetInstance();
 
-            if (shimTextBuffer != null)
+            try
             {
-                foreach (var project in _projectContainer.GetProjects())
+                projectsBuilder.AddRange(_projectContainer.GetProjects());
```

Comment:
Why are we taking these and then stuffing them in the array builder? Aren't they an ImmutableArray we can enumerate over?